src/CodeStyle/Tools/Program.cs

```diff
@@ -0,0 +1,302 @@
+
+                                case AnalysisMode.All:
+                                    // Escalate all rules which can be enabled on build.
+                                    isEnabledInNonDefaultMode = !rule.CustomTags.Contains(s_neverTag);
+                                    break;
+
+                                case AnalysisMode.Minimum:
+                                    // Escalate all highly recommended rules.
+                                    isEnabledInNonDefaultMode = rule.CustomTags.Contains(s_highlyRecommendedTag);
+                                    break;
+
+                                case AnalysisMode.Recommended:
+                                    // Escalate all recommended and highly recommended rules.
+                                    isEnabledInNonDefaultMode = rule.CustomTags.Contains(s_highlyRecommendedTag) || rule.CustomTags.Contains(s_recommendedTag);
+                                    break;
+
+                                case AnalysisMode.Default:
+                                    // Retain the default severity and enabled by default values.
+                                    isEnabledInNonDefaultMode = false;
+                                    break;
+
+                                default:
+                                    throw new NotSupportedException();
+                            }
+
+                            return (isEnabledByDefault: rule.IsEnabledByDefault,
+                                    effectiveSeverity: isEnabledInNonDefaultMode ? DiagnosticSeverity.Warning : rule.DefaultSeverity);
+                        }
+
+                        static string GetRuleSeverity(bool isEnabledByDefault, DiagnosticSeverity defaultSeverity)
+                        {
+                            if (isEnabledByDefault)
+                            {
+                                return GetSeverityString(defaultSeverity);
+                            }
+                            else
+                            {
+                                return GetSeverityString(null);
+                            }
+
+                            static string GetSeverityString(DiagnosticSeverity? severity)
+                            {
+                                if (!severity.HasValue)
+                                {
+                                    return "none";
+                                }
+
+                                return severity.Value switch
+                                {
+                                    DiagnosticSeverity.Error => "error",
+                                    DiagnosticSeverity.Warning => "warning",
+                                    DiagnosticSeverity.Info => "suggestion",
+                                    DiagnosticSeverity.Hidden => "silent",
+                                    _ => throw new NotImplementedException(severity.Value.ToString()),
```

Comment:
This is a duplication from an existing internal extension method. I'm not sure if it's okay to link to a file in the workspaces layer from here?